### PR TITLE
Classify grpc requests properly in the http classifier

### PR DESCRIPTION
**Pull request by adleong**
The gRPC protocol always sets the HTTP response status code to 200 and instead communicates failures in a grpc-status header sent in a TRAILERS frame. Linkerd uses the HTTP response status code to determine if a response is successful, and therefore will consider all gRPC responses successful regardless of their gRPC status code. This means that functionality such as retries and circuit breaking do not function correctly with gRPC traffic.

We update the Http classifier to look for the presence of a `Content-Type: application/grpc` header and use Grpc response classification when it is set.

### CODE REVIEW

File: linkerd/app/core/src/classify.rs
Comment (hawkw) [inline linkerd/app/core/src/classify.rs:369]: should we also be testing the case where the `grpc-status` header is in the response start?

File: linkerd/app/core/src/classify.rs
Comment (hawkw) [inline linkerd/app/core/src/classify.rs:83]: So, one difference between the behavior with `Request::Profile` and `Request::Default` is that in the `Profile` case, we will still use the configured HTTP classifier _if the response begins with a non-`200 OK` status code_ (which would indicate that it will not be terminated by a `grpc-status` (IIRC). In the `Default` case, we will always use the gRPC classifier (and just treat any 5xx HTTP status as an error), which is fine since we don't have a user-defined HTTP classifier. 

This code makes the `ClientPolicy::Http` classifier behave the same as the `Default` classifier, rather than the `Profile` classifier. I'm not sure if this is actually the most desirable behavior here --- would it be more correct to behave more like the `Profile` classifier, and use the configured HTTP classifier if the response has a 5xx HTTP status and there is no `grpc-status` trailer?

Comment (adleong) [reply linkerd/app/core/src/classify.rs]: If the `Content-Type: application/grpc` header is present, I would expect the request to be classified entirely based on `grpc-status` and for the HTTP response status to be ignored, even if it's a non-200.

Comment (hawkw) [reply linkerd/app/core/src/classify.rs]: I also think that this behavior is pretty reasonable; I just wanted to point out that it's different from what we do for ServiceProfiles, and make sure we were okay with that.

File: linkerd/app/core/src/classify.rs
Comment (hawkw) [inline linkerd/app/core/src/classify.rs:86]: nit, take it or leave it: it occurs to me that we don't actually need to repeat this branch twice, and we could instead write
```rust
        match self {
            Self::Profile(classes) => Response::Profile(classes.clone()),
            Self::ClientPolicy(ClientPolicy::Grpc(policy)) => Response::Grpc(policy.clone()),
            // If the request's content type indicates that it is gRPC, use the
            // default gRPC classifier rather than a configured HTTP classifier.
            _ if is_grpc(req.headers()) => Response::Grpc(Default::default()),
            Self::ClientPolicy(ClientPolicy::Http(policy)) => Response::Http(policy.clone()),
            Self::Default => Response::default(),
        }
```
but it's up to you whether you think that's easier to follow or not...

**Review by hawkw — APPROVED** [review]
This looks good to me! I'd like to get @olix0r's approval as well before merging.

**Review by olix0r — APPROVED** [review]

**Pull request merged**
